File: day03/day03.go
package day03

import (
	"strings"
)

type Command struct{}

func (cmd *Command) Execute(input string) (any, any) {
	lines := strings.Split(input, "\n")
	return cmd.part1(lines), cmd.part2(lines)
}

func (cmd *Command) part1(lines []string) int {
	var prioritySum int

	for _, line := range lines {
		bagSize := len(line)
		comp1, comp2 := line[:bagSize/2], line[bagSize/2:]
		common := intersect(comp1, comp2)
		prioritySum += itemPriority(common)
	}

	return prioritySum
}

func (cmd *Command) part2(lines []string) int {
	var prioritySum int

	for i := 0; i < len(lines); i = i + 3 {
		common := intersect(lines[i+2], intersect(lines[i], lines[i+1]))
		prioritySum += itemPriority(common)
	}

	return prioritySum
}

func intersect(a, b string) string {
	intersection := strings.Builder{}
	for _, c := range a {
		if strings.ContainsRune(b, c) && !strings.ContainsRune(intersection.String(), c) {
			intersection.WriteRune(c)
		}
	}
	return intersection.String()
}

func itemPriority(item string) int {
	const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	for _, char := range item {
		for i, c := range alphabet {
			if c == char {
				return i + 1
			}
		}
	}
	return 0
}

File: day03/day03.ts
import { readFileSync } from "../utils/readfile.ts";

const items = [..." abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"];

const parseInput = (path: string): string[] => {
  return readFileSync(path).split("\n");
};

const part1 = (parsedInput: string[]) => {
  const input = parsedInput
    .map((v) => [
      v.slice(0, v.length / 2).split(""),
      v.slice(v.length / 2).split(""),
    ]);

  let prioritySum = 0;
  for (const sack of input) {
    const commonItem = sack[0].find((c) => sack[1].includes(c));
    prioritySum += items.indexOf(commonItem!);
  }

  console.log("Part 1:", prioritySum);
};

const part2 = (parsedInput: string[]) => {
  let prioritySum = 0;
  for (let i = 0; i < parsedInput.length; i += 3) {
    const badgeItem = [...parsedInput[i]].find(
      (item) =>
        parsedInput[i + 1].includes(item) && parsedInput[i + 2].includes(item),
    );
    prioritySum += items.indexOf(badgeItem!);
  }

  console.log("Part 2:", prioritySum);
};

export default function (inputPath: string) {
  const parsedInput = parseInput(inputPath);

  part1(parsedInput);
  part2(parsedInput);
}
